feat(machines): update DHCP and discovery forms to use the new API #1116

petermakowski opened the pull request:
## Done

- update DHCP add form to use list box
- Update discovery form to use the new API and machine list box
- keep previous data to prevent flashes of no data in between data requests
- add e2e test for dhcp snippets
- add `SelectButton` component
- add OutsideClickHandler component

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to `MAAS/r/settings/dhcp`
- Select Type: Machine and pick a machine from the list 
- Go to network discovery
- Expand the dropdown on the right hand side and select "Add discovery..."
- Select a machine from the list in the Parent field and
- Ensure the device.create websocket message has been sent with the correct payload (including the id of a machine which was selected as a parent)
- You can look at websocket messages using developer tools
  - go to network tab -> filter by WS 
  - look for an entry starting with `ws?csrftoken=` 
- You can verify the parent id is valid by visiting the machine details page with a given id and verifying it's the one that was requested in the form (e.g. `/MAAS/r/machine/[replace-this-with-your-selected-parent-id]`)


## Fixes

Fixes: https://github.com/canonical/app-tribe/issues/1116
Fixes: https://github.com/canonical/app-tribe/issues/1112
Fixes: https://github.com/canonical/app-tribe/issues/1340

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against _main_ rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in main.

## Screenshots


![Kapture 2022-09-07 at 15 55 02](https://user-images.githubusercontent.com/7452681/188896348-8b71ee3d-9d84-4ae9-8b2d-e26042bb0c7c.gif)

webteam-app commented:
Demo starting at https://maas-ui-4389.demos.haus

petermakowski commented:
@rayito Decided not to implement the absolute positioning for the select dropdown - it can be tricky when there's not enough content on the page parts of the dropdown may get cut off. If you feel strongly about it I'd suggest we look into this in a separate ticket.

rayito commented:
> 

Totally fine

> @rayito Decided not to implement the absolute positioning for the select dropdown - it can be tricky when there's not enough content on the page parts of the dropdown may get cut off. If you feel strongly about it I'd suggest we look into this in a separate ticket.

Totally fine, closing it by clicking outside was more important than that. UX+1

petermakowski commented:
Searching by machine id doesn't currently work and will be looked at in a separate ticket https://github.com/canonical/app-tribe/issues/1339

huwshimi commented on src/app/base/hooks/base.ts:
```diff
@@ -175,3 +175,24 @@ export const useScrollToTop = (): void => {
   }, [pathname]);
 };
+
+export const usePreviousPersistent = <T extends unknown>(
```
This could do with a docstring and some comments below.

huwshimi replied:
I think it's worth adding these docstrings as it's surprising how easy it is to come back and not remember what's doing what.

I think some comments inside the hook would be worth adding as well to help follow the logic.

huwshimi commented on src/app/store/machine/utils/hooks.ts:
```diff
@@ -181,6 +182,7 @@ export const useFetchMachine = (
   const machine = useSelector((state: RootState) =>
     machineSelectors.getById(state, id)
   );
+  const previousMachine = usePreviousPersistent(machine);
```
I think the machine select is the only case we have right now that needs the previous data. I wonder if we should move this to `MachineSelect` until we have more than one use case? Not really sure, it's probably not that costly to compute this, it just feels like it fits this use case, but might not fit all cases (it's not really keeping the previous data, it's storing a local copy, but the data will be removed from the store so can't be selected).

petermakowski replied:
I am fairly sure it will be useful for other use cases in the future, but went with your suggestion and moved this to `MachineSelect` until that actually happens.

huwshimi reviewed (changes requested):
Hey @petermakowski apologies for the "code -1" but I think there are big enough changes needed that this will require another review.

Happy to chat if any of my comments don't make sense.

petermakowski commented on src/app/base/hooks/base.ts:
```diff
@@ -175,3 +175,24 @@ export const useScrollToTop = (): void => {
+
+export const usePreviousPersistent = <T extends unknown>(
+  value: T,
+  isEqualFn?: (prev: T, next: T) => boolean
```
Might be useful in the future, but can remove for now.

petermakowski commented on src/app/base/components/DhcpFormFields/MachineSelect/MachineSelect.tsx:
```diff
@@ -0,0 +1,80 @@
+    dispatch(tagActions.fetch());
+  }, [dispatch]);
+
+  return (
```
Absolute positioning isn't related to whether we use a portal or not, it's more about the height of the element that can potentially exceed the height of the page and be clipped.

Agree it would be a good idea for the two to share a common base.

petermakowski replied:
Will follow up separately in https://github.com/canonical/maas-ui/issues/4397

petermakowski commented on src/app/base/components/OutsideClickHandler/OutsideClickHandler.tsx:
```diff
@@ -0,0 +1,24 @@
+import { useEffect, useRef } from "react";
+
+const OutsideClickHandler = ({
```
Oh nice, didn't realise. Will refactor this component to use it.

petermakowski replied:
This will be resolved in https://github.com/canonical/maas-ui/issues/4397

huwshimi replied:
I think this worth resolving in this PR as there's no need to add this component and then remove it.

petermakowski replied:
I actually already did this within this PR.

huwshimi replied:
Ah cool, I think you can probably do away with this component and just use the hook directly inside `MachineSelect`.

petermakowski commented on src/app/base/components/DhcpFormFields/MachineSelect/MachineSelectBox.tsx:
```diff
@@ -0,0 +1,53 @@
+        searchText={searchText}
+        setSearchText={setSearchText}
+      />
+      <Pagination
```
Will followup on in https://github.com/canonical/maas-ui/issues/4396

petermakowski commented on src/app/base/components/DhcpFormFields/DhcpFormFields.tsx:
```diff
@@ -134,21 +131,30 @@ export const DhcpFormFields = ({ editing }: Props): JSX.Element => {
       />
-      {type &&
+      {type === "machine" ? (
+        <MachineSelect
```
This will be followed up in https://github.com/canonical/maas-ui/issues/4397

huwshimi replied:
I think this is worth doing in this PR as it doesn't relate to `ContextualMenu`.

petermakowski replied:
I am happy to do that, but would you mind if I did this separately in https://github.com/canonical/app-tribe/issues/1340.

I prefer to move in small increments in each PR, building on top of changes from previous work. This can be very helpful especially in case of us working in different time zones and the lag in code reviews.

huwshimi commented on src/app/base/components/DebounceSearchBox/DebounceSearchBox.tsx:
```diff
@@ -22,6 +22,8 @@ const DebounceSearchBox = ({
   searchText,
   setSearchText,
+  autoFocus,
+  placeholder,
```
It looks like the type includes the `SearchBoxProps` but we don't spread them, which we could do instead of passing these specific props.

huwshimi commented on src/app/base/components/DhcpFormFields/DhcpFormFields.test.tsx:
```diff
@@ -9,24 +10,51 @@ import { Labels } from "./DhcpFormFields";
 const mockStore = configureStore();
-
+const machines = [
+  machineFactory({
```
Do these tests need all these specific props defined? If not it'd probably be clearer to just define the ones we need (if any).

huwshimi commented on src/app/base/components/DhcpFormFields/MachineSelect/MachineSelect.tsx:
```diff
@@ -0,0 +1,80 @@
+  selected?: Machine["system_id"] | null;
+};
+
+export const MachineSelect = ({
```
I don't think this is a part of sharing a base with ContextualMenu as ContextualMenu is not a field. I think it's probably worth doing the field refactor as part of this PR as it's a significant rework of how a lot of this is implemented.

petermakowski replied:
Fair enough, created a separate issue: https://github.com/canonical/app-tribe/issues/1340

huwshimi reviewed (changes requested):
Thanks for doing those changes @petermakowski. I think changing this to a `FormikField` is worth doing in this PR (but not the ContextualMenu parts) as a fair bit of this will be made redundant.

huwshimi commented on src/app/base/components/DhcpFormFields/MachineSelect/MachineSelectBox/MachineSelectBox.tsx:
```diff
@@ -0,0 +1,55 @@
+  onSelect: (machine: Machine | null) => void;
+}): JSX.Element => {
+  const [searchText, setSearchText] = useState("");
+  const [deboucedText, setDebouncedText] = useState("");
```
Just noticed this says `deboucedText`.

petermakowski replied:
Oops! Fixed.

petermakowski commented:
@huwshimi  I addressed all your comments except the ones that I moved to separate issues (https://github.com/canonical/app-tribe/issues/1340, https://github.com/canonical/maas-ui/issues/4397, https://github.com/canonical/maas-ui/issues/4396) and this is ready for another review now.

petermakowski commented on src/app/base/components/DhcpFormFields/MachineSelect/MachineSelect.tsx:
```diff
@@ -0,0 +1,100 @@
+    <div className="machine-select">
+      {/* TODO: update once Field allows a custom label id
+      https://github.com/canonical/react-components/issues/820 */}
+      <Field label={<span id={labelId}>{label}</span>} {...props}>
+        <OutsideClickHandler onClick={() => setIsOpen(false)}>
+          <SelectButton
+            aria-describedby={labelId}
```
This is in order to get additional description for the button element, rendered result below.
<img width="349" alt="image" src="https://user-images.githubusercontent.com/7452681/189324020-e03565d6-8b3b-4c99-98ed-456ffd0126e2.png">

petermakowski commented:
@huwshimi could you take another look at this please?

ndv99 reviewed (commented):
You can probably remove `Loading` from this enum since it isn't used in the file any more
https://github.com/canonical/maas-ui/pull/4389/files#diff-3fc2d13129aa52e7573026eac1c141e58e9f922cf86d851e9e56fd4902f85ba5R19-R23

ndv99 reviewed (commented):
Happy with the code over-all but obviously QA can't be done until that error message in the discovery form is dealt with

petermakowski commented:
> Happy with the code over all but obviously QA can't be done until that error message in the discovery form is dealt with

As it's due to a bug in MAAS I think it's fine to QA by verifying the right websocket message is being sent. Updated the QA description.

ndv99 reviewed (approved):
LGTM!
